### ISSUE TITLE
SSL Weight Decay

### ISSUE THREAD

**Issue by guarin** [microsoft/torchgeo#1828]
### Summary

I noticed that your SSL models use weight decay for all their parameters, see for example:

https://github.com/microsoft/torchgeo/blob/0f8b0ac3eabf1e4156e7259d34b1b63e6df5d445/torchgeo/trainers/simclr.py#L280-L284

For these models it is usually better to not use weight decay for batch norm and bias parameters. We implement it like this in lightly:

https://github.com/lightly-ai/lightly/blob/c1e341222b2d199f8a23277a92417debf8caa783/benchmarks/imagenet/resnet50/simclr.py#L65-L84

Could give a small boost to your SSL models :)

**Comment by adamjstewart** [comment]
Thanks for the suggestion! We didn't have a lot of prior experience with SSL so we chose to match the defaults of the original SimCLR/MoCo papers. Do you know of any papers that demonstrate that no weight decay works better? I'm surprised Google/FAIR didn't find this during their hyperparameter tuning.

**Comment by guarin** [comment]
I don't think it is mentioned in the SimCLR paper but it is in the code here: https://github.com/google-research/simclr/blob/383d4143fd8cf7879ae10f1046a9baeb753ff438/tf2/model.py#L40-L42

BYOL does the same: https://github.com/google-deepmind/deepmind-research/blob/f5de0ede8430809180254ee957abf36ed62579ef/byol/byol_experiment.py#L191-L195

But I just noticed that you are not using LARS optimizer and in SimCLR they only did this for LARS. For the other optimizers they didn't use weight decay at all, but I am not sure if they benchmarked their code with these settings.

**Comment by adamjstewart** [comment]
Okay, finally had time to look into this.

### SimCLR

> I don't think it is mentioned in the SimCLR paper

Weight decay is mentioned in:

* Section 2.3 of [SimCLR v1](https://arxiv.org/abs/2002.05709)
* Section 3.1 of [SimCLR v2](https://arxiv.org/abs/2006.10029).

> For the other optimizers they didn't use weight decay at all

You are correct that weight decay is not used in the [optimizer](https://github.com/google-research/simclr/blob/383d4143fd8cf7879ae10f1046a9baeb753ff438/model_util.py#L86), although it is used in the [loss function](https://github.com/google-research/simclr/blob/383d4143fd8cf7879ae10f1046a9baeb753ff438/model.py#L92).

### MoCo

Weight decay is mentioned in:

* Section 4 of [MoCo v1](https://arxiv.org/abs/1911.05722)
* Section 5 of [MoCo v3](https://arxiv.org/abs/2104.02057)

It isn't mentioned in MoCo v2, although the code for v2 is largely the same as v1. The value of weight decay for v3 is not mentioned in the paper, just that it was used.

In the code base, weight decay is used with [SGD](https://github.com/facebookresearch/moco/blob/main/main_moco.py#L291) in v1/v2, [LARS](https://github.com/facebookresearch/moco-v3/blob/c349e6e24f40d3fedb22d973f92defa4cedf37a7/main_moco.py#L225) in v3, and [AdamW](https://github.com/facebookresearch/moco-v3/blob/c349e6e24f40d3fedb22d973f92defa4cedf37a7/main_moco.py#L229) in v3.

**Comment by adamjstewart** [comment]
If you want to submit a PR that removes weight decay from our SimCLR optimizer and adds it to our loss function, I would be happy to accept it. I'm a little afraid to remove it entirely though.

**Comment by adamjstewart** [comment]
I think this issue can be closed. If users want to reproduce the original MoCo/SimCLR papers, they can use our current defaults. If they want to try to improve performance, they can use `weight_decay=0`.
